### PR TITLE
Build fixes, version 3

### PR DESCRIPTION
**Pull request by ocrete**
The current code doesn't actually compile!

I also fixed a number of things to make the installation compliant with the Linux File Hierarchy Standard, so that the files are where they are expected.

I also fixed the pkgconfig file to be more compliant.

### CODE REVIEW

**Comment by jakeolenick** [comment]
Apologies, we don't yet have a formal process for reviewing external PRs yet. I also don't have a company github so, I'm using my personal one from when I was in school. In any case, this is Jacob Olenick (Jake) from V-Nova, I'll try to get this reviewed by the end of the day

File: CMakeLists.txt [hunk 1]
```diff
@@ -30,7 +30,7 @@ option(VN_SDK_USE_GHC_FILESYSTEM "Use ghc-filesystem instead of C++ standard <fi
 option(VN_SDK_COVERAGE "Generate gcov statistics for coverage" OFF)
 
 set(VN_SDK_FFMPEG_LIBS_PACKAGE
-    "ffmpeg-libs-minimal"
+    "pkgconfig"
```
Comment (jakeolenick) [inline CMakeLists.txt:33]: I'd suggest making this a change to the in `docs/building_linux.md`, rather than changing the default directly. In particular, this change is equivalent to adding `-DVN_SDK_FFMPEG_LIBS_PACKAGE=pkgconfig` to the `cmake -G ..` step

It sounds like you've attempted to build completely without _any_ use of conan, which has never been done before as far as I'm aware. So, any contributions to `building_linux.md` and `getting_started_linux.md` would probably be quite valuable.

File: src/core/decoder/src/decode/huffman.c [hunk 2]
```diff
@@ -21,15 +21,15 @@
 
 /*- General utility functions -------------------------------------------------------------------*/
 
-static inline uint8_t clz(int32_t streamData, uint8_t numBits)
+static inline uint8_t clz(uint32_t streamData, uint8_t numBits)
```
Comment (jakeolenick) [inline src/core/decoder/src/decode/huffman.c:24]: Why this change? We're using `clz` on `uint8_t` data generally, and all three `clz` functions expect `int32_t`s I thought?

File: src/core/decoder/src/decode/huffman.c [hunk 3]
```diff
@@ -21,15 +21,15 @@
...
 {
 #if defined WIN32
     return (uint8_t)(_lzcnt_u32(streamData) + numBits - 32);
 #elif VN_COMPILER(GCC) && (__GNUC__ >= 14)
     /* Annoyingly, it's undefined behaviour if you provide 0 to __builtin_clz. To match the Windows
      * behaviour (where 0 is just another leading zero), we need clzg, with sizeof(streamData) as
      * the default arg. */
-    return (uint8_t)(__builtin_clzg(streamData, (sizeof(streamData) * 8)) + numBits - 32);
+    return (uint8_t)(__builtin_clzg(streamData, (int) (sizeof(streamData) * 8)) + numBits - 32);
```
Comment (jakeolenick) [inline src/core/decoder/src/decode/huffman.c:32]: Maybe `int32_t` rather than `int`?

Also, looking at this, I wonder if it's better to provide `numBits` (or indeed `(int32_t) numBits`) directly, as the second argument. That would make it more obviously identical to the pre-gcc-14 code..

File: src/core/decoder/src/decode/huffman.c (hunk 2 above)
Comment (jakeolenick) [reply src/core/decoder/src/decode/huffman.c:24]: Ah wait no, you're right, they all expect `uint32_t`s, good call.

File: src/utility/src/math_utils.h [hunk 4]
```diff
@@ -32,7 +32,7 @@ static inline uint8_t clz(uint32_t n)
 #if defined WIN32
     return (uint8_t)(_lzcnt_u32(n));
 #elif defined __linux__ && (__GNUC__ >= 14)
-    return (uint8_t)(__builtin_clzg(n, (8 * sizeof(n)));
+    return (uint8_t)(__builtin_clzg(n, (int) (8 * sizeof(n))));
```
Comment (jakeolenick) [inline src/utility/src/math_utils.h:35]: Likewise, probably better as `int32_t`

File: src/utility/src/base_decoder_libav.cpp [hunk 5]
```diff
@@ -273,7 +274,7 @@ int BaseDecoderLibAV::addFilter(std::string_view filter)
 
     const std::string args =
         fmt::format("width={}:height={}:pix_fmt={}:time_base=1/1:sar={}/{}", m_videoDecCtx->width,
-                    m_videoDecCtx->height, m_videoDecCtx->pix_fmt,
+                    m_videoDecCtx->height, (int) m_videoDecCtx->pix_fmt,
```
Comment (jakeolenick) [inline src/utility/src/base_decoder_libav.cpp:277]: Why this change? Are you getting a compiler warning for providing an enum, or does it just format enums weirdly?

File: CMakeLists.txt [hunk 6]
```diff
@@ -124,21 +124,25 @@ if(VN_SDK_DOCS)
     lcevc_add_subdirectory(docs/sphinx)
 endif()
 
+include(GNUInstallDirs)
+
 # Record configuration - header and README
```
Comment (jakeolenick) [inline CMakeLists.txt:130]: Update comment to mention/explain the new configuration file?

File: CMakeLists.txt [hunk 7]
```diff
@@ -124,21 +124,25 @@ if(VN_SDK_DOCS)
...
 #
-install(FILES "COPYING" DESTINATION "./")
-install(FILES "${CMAKE_BINARY_DIR}/install_config/README.md" DESTINATION "./")
-install(FILES "cmake/install/conanfile.txt" DESTINATION "./")
-install(FILES "cmake/install/lcevc_dec.pc" DESTINATION "./pkgconfig/")
-install(DIRECTORY "licenses" DESTINATION "./")
-install(FILES "LICENSE.md" DESTINATION "./licenses/LCEVCdec")
+install(FILES "COPYING" DESTINATION "${CMAKE_INSTALL_DOCDIR}")
+install(FILES "${CMAKE_BINARY_DIR}/install_config/README.md" DESTINATION "${CMAKE_INSTALL_DOCDIR}")
+install(FILES "cmake/install/conanfile.txt" DESTINATION "${CMAKE_INSTALL_DOCDIR}")
```
Comment (jakeolenick) [inline CMakeLists.txt:140]: I think conanfile.txt probably belongs in `${CMAKE_INSTALL_LIBDIR}` since that's the same place as the equivalent `pkgconfig` file.

File: src/sample_cpp/CMakeLists.txt [hunk 8]
```diff
@@ -24,10 +24,12 @@ add_executable(lcevc_dec::sample ALIAS lcevc_dec_sample)
 install(TARGETS "lcevc_dec_sample" DESTINATION "bin")
 
 if(VN_SDK_SAMPLE_SOURCE)
-    install(FILES ${SOURCES} DESTINATION "sample_cpp")
-    install(FILES "standalone/README.md" DESTINATION "sample_cpp")
-    install(FILES "standalone/CMakeLists.txt" DESTINATION "sample_cpp")
-    install(FILES "standalone/conanfile.py" DESTINATION "sample_cpp")
-    install(FILES "standalone/requirements.txt" DESTINATION "sample_cpp")
-    install(DIRECTORY "standalone/cmake" DESTINATION "sample_cpp")
+   include(GNUInstallDirs)
```
Comment (jakeolenick) [inline src/sample_cpp/CMakeLists.txt]:
```suggestion
    include(GNUInstallDirs)
```

Very recently, we've introduced formatting (including cmake formatting) as a pre-commit git hook, so this would now get caught automatically.

**Comment by jakeolenick** [comment]
Alright, that's all from me.

The introduction of `GNUIncludeDirs` is very handy! I didn't know about that previously.

The changes here MAY undergo some further changes, if they don't pass our internal CI (we don't yet have a public CI). However, I don't think this will need many (or any) changes.

Also, needless to say, it's not strictly true to say that "The current code doesn't actually compile!". You just have to follow the steps in `getting_started_linux.md` and `building_linux.md` 😛

File: src/core/decoder/src/decode/huffman.c (hunk 3 above)
Comment (ocrete) [reply src/core/decoder/src/decode/huffman.c:32]: From reading https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html#index-_005f_005fbuiltin_005fclzg and https://clang.llvm.org/docs/LanguageExtensions.html#builtin-clzg-and-builtin-ctzg  it says the second argument is specifically an int.

File: src/utility/src/math_utils.h (hunk 4 above)
Comment (ocrete) [reply src/utility/src/math_utils.h:35]: Same as above, the 2nd argument is document as an int

File: src/utility/src/base_decoder_libav.cpp (hunk 5 above)
Comment (ocrete) [reply src/utility/src/base_decoder_libav.cpp:277]: I'm getting an error when compiling, because three is no formatter for the pix_fmt enum

File: CMakeLists.txt (hunk 7 above)
Comment (ocrete) [reply CMakeLists.txt:140]: Is there a specific place to install conan files?

File: CMakeLists.txt (hunk 1 above)
Comment (ocrete) [reply CMakeLists.txt:33]: The reliance of Conan won't work if you want the library distributed as part of the various Linux distros, with these patches, I can compile directly on Fedora 20 using GCC 14.1.1

File: CMakeLists.txt (hunk 6 above)
Comment (ocrete) [reply CMakeLists.txt:130]: It's not really a configuratioin file, just a set of defaults based on the prefix?

File: CMakeLists.txt (hunk 7 above)
Comment (ocrete) [reply CMakeLists.txt:140]: I'm not sure I understand which is the case where you need to use the conanfile with the binaries, but not using lcevcdec as a Conan package itself ?

File: src/core/decoder/src/decode/huffman.c (hunk 3 above)
Comment (jakeolenick) [reply src/core/decoder/src/decode/huffman.c:32]: Fair, may as well keep it consistent.

File: src/utility/src/base_decoder_libav.cpp (hunk 5 above)
Comment (jakeolenick) [reply src/utility/src/base_decoder_libav.cpp:277]: Ah interesting! It's possible that more recent versions are more picky about this (I assume that when we build, it's just converting the enum to the underlying type)

In any case, we prefer `int32_t` unless there's a compelling reason not to use it

File: CMakeLists.txt (hunk 6 above)
Comment (jakeolenick) [reply CMakeLists.txt:130]: Fair. Possibly a better comment might be:

`# Generate the config header, README, and pkgconfig defaults`

File: CMakeLists.txt (hunk 1 above)
Comment (jakeolenick) [reply CMakeLists.txt:33]: Hm... You may be right that this is a better default.

There will probably be some extremely annoying downstream effects on our CI (which does currently rely on Conan) and our build instructions. But... Technically yeah, probably better...

File: CMakeLists.txt (hunk 7 above)
Comment (jakeolenick) [reply CMakeLists.txt:140]: Hm... I might be mistaken about what these two files are for. Is `lcevc_dec.pc` not simply the pkgconfig equivalent of `conanfile.txt`? That is, they're both files which allow you to produce a package of `LCEVCdec`?

I'm not sure I understand the relevance of your questions...

Comment (ocrete) [reply CMakeLists.txt:140]: The .pc file is the information needed to build/link against the LCEVCdec library. It's particularly needed if you link statically as in this case you need to link against multiple libraries in the right order. It also includes dependencies. on other libraries, etc.

Comment (jakeolenick) [reply CMakeLists.txt:140]: Hm, that's what I was thinking. In that case, it does sound like `lcevc_dec.pc` and `conanfile.txt` serve the same purpose (for different package managers), and therefore they should either _both_ go in `LIBDIR`, or else _neither_ belong in `LIBDIR`?

(You've got `lcevc_dec.pc` in `LIBDIR`, while `conanfile.txt` is in `DOCDIR`)

Comment (ocrete) [reply CMakeLists.txt:140]: The libdir location is baked into the pkg-config tool.

I couldn't find any doc on where conan expects it?

Comment (jakeolenick) [reply CMakeLists.txt:140]: Ahh now I get what you mean. I _think_ conanfile.txt can go anywhere. That is, I think you always have to provide a path to its location, so there's no default (I think)

**Comment by jakeolenick** [comment]
Basically just the indentation, the int32_t, the comment, and (optionally) the location for conanfile.txt, and then this is good to go.

File: src/utility/src/base_decoder_libav.cpp (hunk 5 above)
Comment (ocrete) [reply src/utility/src/base_decoder_libav.cpp:277]: The reason I chose int is that pix_fmt is a "enum AVPixelFormat", and in the C language specification, the guarantee is that enum is the same size as int, so by casting to int, we'll be sure it will work on all platform. Although I guess any platform where int!=int32_t will probably break somewhere else.

**Comment by ocrete** [comment]
I pushed an update with some changes.

**Comment by ocrete** [comment]
My logic to leaving the conanfile.txt in the doc dir is that if you are a naive packager, you can just delete the docs dir and the include dirs, and the software will still work, while stuff in the rest of /usr/share or /usr/lib must be kept for the proper functionality of the library

**Comment by jakeolenick** [comment]
Alright, the only remaining stuff is cosmetic so, I think this is good to go in

(Sorry again for delay, just realised updates were going to my private email rather than work)
